Begin instruction for differential expression notebook

allyhawkins opened the pull request:
This PR begins to outline the code and structure for the notebook on differential expression. I mostly took from the notebook that I used to explore the dataset that we were using, following a lot of the same steps but just with DESeq. 

The general structure of the notebook that I setup is: 
1. exploring and setting up the dataset 
2. discussion of pseudo-bulking 
3. DESeq2 
4. evaluating the results from DESeq2 and making some plots 
5. Comparing abundances (does this still make sense in this notebook?)

There are some key things to note while reviewing: 

- The integrated dataset that is being loaded in this notebook was setup following the snakefile in #589, where output from ScPCA (`_filtered.rds` files) were filtered following the setup in the `prepare_rms.R` script and then integrated using `fastMNN`. The integrated dataset used in the exploratory notebook varied slightly in that it came from when we integrated with our benchmarking workflow. The main difference here is that those `_filtered.rds` files were first run through downstream analyses and were filtered following those guidelines. This included removal of genes that were [detected in < 5% of cells and a mean gene expression of < 0.1](https://github.com/AlexsLemonade/scpca-downstream-analyses/blob/ccb93eead4e57d8940100e247cbbe14614081fdd/core-analysis/01-filter-sce.R#L245-L249). Because of this the number of starting genes was less and the integrated UMAPs do appear slightly different. Now, with the inclusion of lowly covered genes, cells from ARMS are less likely to integrate with cells from ERMS patients, even if they are from the same cell type. The differences in gene expression between these patients is well known and also part of the premise of the paper for this dataset, so I think this is okay, but I want reviewers to be aware of this when looking at the "library selection" and dataset exploration part of the notebook. 
- I left italics throughout where I would still want to expand on the text, but generally noting that I don't want to spend a lot of time on the text here, mostly on the code. Some of the text here is copied from the setup notebook, but we should break up cleaning up each section in subsequent PRs. My main focus is on the code and then flow and order of concepts.
- I only included walking through a single comparison of one cell type and thought that it was getting long so did not include a second cell type right now. I think that may be something we want to move to the exercises? 
- I also did not complete the section on comparing abundances yet, mostly because it was getting long, but also because I kind of feel like that might be a separate notebook? 

I've included a rendered version of the notebook for reference: 
[04-differential_expression.html.zip](https://github.com/AlexsLemonade/training-modules/files/10230800/04-differential_expression.html.zip)

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -0,0 +1,544 @@
+Pseudo-bulking creates a new counts matrix that contains the total counts across all cells of a given label (e.g. cell type) for each sample. 
+This allows us to use single-cell resolution to define the labels and sum gene counts across groups of cells containing the same label, and avoids counting each cell as its own replicate. 
+
+Pseudo-bulking is implemented prior to differential expression analysis for the following reasons: 
```
```suggestion
Pseudo-bulking is implemented prior to differential expression analysis on single-cell data for the following reasons: 
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -0,0 +1,544 @@
+
+Pseudo-bulking is implemented prior to differential expression analysis for the following reasons: 
+
+- Produces larger counts, which are more amenable to standard normalization and differential expression methods used by bulk RNA-sequencing. 
```
up to you - 
```suggestion
- Produces larger counts, which allows us to use standard normalization and differential expression methods used by bulk RNA-sequencing. 
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -0,0 +1,544 @@
+
+- Produces larger counts, which are more amenable to standard normalization and differential expression methods used by bulk RNA-sequencing. 
+-  Collapses gene expression counts by sample, so that samples, rather than cells, represent replicates.
+- Masks variance within a sample to highlight variance across samples.
```
```suggestion
- Masks variance within a sample to emphasize variance across samples.
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -0,0 +1,544 @@
+-  Collapses gene expression counts by sample, so that samples, rather than cells, represent replicates.
+- Masks variance within a sample to highlight variance across samples.
+
+We are staring with the simple comparison of looking at one cell type and comparing across two groups of samples, samples belonging to either the ERMS and ARMS subtypes. 
```
```suggestion
We are starting with the simple comparison of looking at one cell type and comparing across two groups of samples: the ERMS and ARMS subdiagnoses. 
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -0,0 +1,544 @@
+- Masks variance within a sample to highlight variance across samples.
+
+We are staring with the simple comparison of looking at one cell type and comparing across two groups of samples, samples belonging to either the ERMS and ARMS subtypes. 
+When we do this, we will first want to pseudo-bulk our dataset to group our cells by cell type and by sample.
```
```suggestion
Our first step will be to pseudo-bulk our dataset to group our cells by cell type and by sample.
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -0,0 +1,544 @@
+
+We are staring with the simple comparison of looking at one cell type and comparing across two groups of samples, samples belonging to either the ERMS and ARMS subtypes. 
+When we do this, we will first want to pseudo-bulk our dataset to group our cells by cell type and by sample.
+This will create a new `SingleCellExperiment` object that contains the pseudo-bulked counts to be used as input for differential expression analysis. 
+Then we can subset the pseudo-bulked object to just contain our cell type of interest before proceeding. 
```
```suggestion
This will create a new `SingleCellExperiment` (SCE) object that contains the pseudo-bulked counts for all genes. 
We can subset this SCE to just contain our cell type of interest (tumor myoblasts) for input to differential expression analysis. 
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -0,0 +1,544 @@
+This will create a new `SingleCellExperiment` object that contains the pseudo-bulked counts to be used as input for differential expression analysis. 
+Then we can subset the pseudo-bulked object to just contain our cell type of interest before proceeding. 
+
+Before we apply pseudo-bulking to our dataset, lets look at a simple example of how pseudo-bulking works. 
```
```suggestion
Before we apply pseudo-bulking to our dataset, let's look at a simple example of how pseudo-bulking works. 
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -0,0 +1,544 @@
+```{r}
+# create new matrix, summing the counts for each gene across all cells in the specified group
+pseudobulked_counts <- cbind(
+  rowSums(counts_mtx[,1:2]), # sum all counts for cells labeled with group A
````
a bit more explicit - 
```suggestion
  rowSums(counts_mtx[,c("A-cell1", "A-cell2")]), # sum all counts for cells labeled with group A
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -0,0 +1,544 @@
+# create new matrix, summing the counts for each gene across all cells in the specified group
+pseudobulked_counts <- cbind(
+  rowSums(counts_mtx[,1:2]), # sum all counts for cells labeled with group A
+  rowSums(counts_mtx[,3:4]) # sum all counts for cells labeled with group B
```
```suggestion
  rowSums(counts_mtx[,c("B-cell1", "B-cell2")]) # sum all counts for cells labeled with group B
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -0,0 +1,544 @@
+pseudobulked_counts 
+```
+Now the actual pseudo-bulking for our dataset! 
+We can pseudo-bulk using any groups that we are interested in.
````
```suggestion
We can pseudo-bulk using any grouping that we are interested in.
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -0,0 +1,544 @@
+```
+Now the actual pseudo-bulking for our dataset! 
+We can pseudo-bulk using any groups that we are interested in.
+For right now, we would are interested in looking at gene expression across cell types, so we want to group the counts matrix by both cell type and sample. 
````
```suggestion
For right now, we would are interested in looking at gene expression across cell types, so we want to group the pseudo-bulked counts matrix by both cell type and sample. 
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -0,0 +1,544 @@
+For right now, we would are interested in looking at gene expression across cell types, so we want to group the counts matrix by both cell type and sample. 
+
+```{r}
+# first let's subset our coldata to only have the columns we care about in pseudobulking 
````
```suggestion
# first let's subset our coldata to only have the columns we care about in pseudo-bulking 
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -0,0 +1,544 @@
+# first let's subset our coldata to only have the columns we care about in pseudobulking 
+pseudobulked_coldata <- colData(rms_sce)[, c("celltype_broad", "sample")]
+
+# this creates a new SCE object that contains the pseudobulked counts across the provided groups 
```
```suggestion
# this creates a new SCE object that contains the pseudo-bulked counts across the provided groups 
```

sjspielman replied:
Small note, we'll want to really emphasize that the _counts_ get summed, not the logcounts (by default).

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -0,0 +1,544 @@
+
+```{r}
+# filter by number of cells 
+filtered_pseudobulked_sce <- filtered_pseudobulked_sce[, pseudobulked_sce$ncells >= 10]
````
Might want to talk about "why 10" here

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -0,0 +1,544 @@
+*Talk about how this specific function uses the sample groups provided to determine the minimum sample size that genes are expressed in.*
+
+```{r}
+# filter out genes that might be lowly expressed using filterbyExpr in `edgeR`
````
might want to note here that we're not actually using edgeR for the differential expression but it has useful utility functions

sjspielman reviewed (commented):
I've reviewed the first half or so, looks like a really nice start! Will come back for more tomorrow but wanted to get this in sooner.

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -0,0 +1,544 @@
+                                               design = ~ diagnosis_groups)
+```
+
+Use the median of ratios method for count normalization followed by regularized log transformation.
````
Again we probably want to reiterate here that we have to normalize "again" because we summed up the raw counts and now have "new" raw counts to normalize. The previous single-cell-resolution logcounts are no longer applicable in this analysis.

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -0,0 +1,544 @@
+```
+
+Evaluate QC here using PCA. 
+Here we can label by the diagnosis groups to show that the highest amount of variance is due to different diagnosis types. 
````
```suggestion
Here we can label by the diagnosis groups to show that the highest amount of variance is due to different subdiagnoses. 
```

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -0,0 +1,544 @@
+colData(integrated_sce) <- DataFrame(combined_coldata_df, row.names = combined_coldata_df$cell_id)
+```
+
+We will pick a gene and look at the expression of that gene across disease type, sample, and cell type.
````
```suggestion
We will pick a gene and look at the expression of that gene across subdiagnosis, sample, and cell type.
```

sjspielman replied:
In next PR(s), might want add some biological context about this gene and RMS.

allyhawkins commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -0,0 +1,544 @@
+
+## Set Up
+
+```{r}
````
Noted that this will be added later in https://github.com/AlexsLemonade/training-modules/issues/573#issuecomment-1353398710.

allyhawkins commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -0,0 +1,544 @@
+
+We can use these plots to help explore all of the samples and visualize the distribution of cell types across our dataset.
+*Need to explain the difference between tumor/tumor_myoblast, etc*
+These plots show that ARMS and ERMS share a lot of the same cell types, but perhaps there are differences in the cells in each sub type.
```
Just noting in general that I plan on re-wording/ updating text in subsequent PRs for each section as noted in https://github.com/AlexsLemonade/training-modules/issues/573#issuecomment-1353398710. So not ignoring these comments, just saving for later!

jashapiro reviewed (commented):
This is a great start.

I do think you will want to comment some more about the fact that the cell types don't cluster together so much. I think we may say that this does mean we will expect to see many differences, but also note that we are not using the directly clustering to inform the DE analysis itself.

(I'm somewhat surprised that the change you see was the result of filtering in downstream analysis, as the integration and clustering used HVGs which shouldn't be any that were filtered out, but obciously I haven't looked into it in much depth!)

I think doing one cell type as the focus is fine. Doing more comparisons across cell types does seem like something we could do in the exercises.

Other than that, I had a number of smaller comments throughout. Let me know if any of those are unclear, or if you just disagree!

allyhawkins commented:
> (I'm somewhat surprised that the change you see was the result of filtering in downstream analysis, as the integration and clustering used HVGs which shouldn't be any that were filtered out, but obciously I haven't looked into it in much depth!)

I just want to quickly respond to this comment. The issue is the initial objects that we are starting with are smaller and genes have been removed when running through downstream analyses so when they get merged, and the merged object only contains genes that are found in all datasets, we have a much smaller object. We then identified HVG on that merged object and there were only ~500 HVG's identified. In contrast, without filtering the merged object now identifies the full set of 2000 HVG's. So the set of genes being used for integration is different with the filtering.

jashapiro commented:
> the merged object only contains genes that are found in all datasets

Ah, yes, that makes sense. Another reason not to filter to expressed genes only on an individual sample basis!

Also a caveat to note in the integration section! `intersect` of the genes may exclude genes you really care about if your individual SCEs were filtered! I am now thinking we probably want to expand that discussion a bit more @sjspielman.

allyhawkins commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -0,0 +1,544 @@
+
+```{r}
+# filter to only the HVG list
+hv_genes <- metadata(integrated_sce)$combined_hvg
````
Yea I wasn't entirely sure about this one... Looking at the OSCA chapter they show the DE analysis example on an object that has been merged and corrected and only contains the HVGs. But the way they choose HVGs is slightly different, instead of using `getTopHVGs` they retain anything with a positive biological component of the variance. So they retain a lot more than 2000 genes. 
Here's the link to how they set up the data that they used for reference: http://bioconductor.org/books/3.16/OSCA.multisample/chimeric-mouse-embryo-10x-genomics.html#chimeric-mouse-embryo-10x-genomics

I think filtering to only the 2000 genes may be too stringent, but I think we definitely want to do some sort of filtering.

jashapiro replied:
I might try it without filtering... I think the main reason for filtering is to speed calculations, but I don't expect that to be too bad.

allyhawkins commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -0,0 +1,544 @@
+shrink_results <- DESeq2::lfcShrink(deseq_object, res = deseq_results, coef = 2, type = "apeglm")
+```
+
+```{r}
+# look at what actually happened when we applyed shrinkage (adjustment of large fold changes that are overestimated)
+comparison_df <- data.frame(
+  lfc_original = deseq_results$log2FoldChange,
+  lfc_shrunken = shrink_results$log2FoldChange,
+  logmean = log10(deseq_results$baseMean)
+  )
+
+ggplot(comparison_df, 
+       aes(x = lfc_original, 
+           y = lfc_shrunken, 
+           color = logmean)) +
+  geom_point(alpha = 0.1) +
+  theme_bw() +
+  scale_color_viridis_c() + 
+  coord_cartesian(xlim = c(-10,10), ylim = c(-10,10))
+```
+
+Not a ton of shrinkage, but there is some. 
````
I am on board with cutting it. To confirm you mean not doing the shrinkage at all? I don't think we need it if it isn't really doing much.

jashapiro replied:
I'd still do the shrinkage, as it is a good thing to do. I'd just make a brief comment about why (don't want to overestimate LFC when expression is low).

jashapiro commented:
One other note: please add this notebook to `scripts/render-live.sh` so we can debug test rendering if needed!

allyhawkins commented:
Thanks all for your helpful feedback here. I've made a number of changes to the notebook and included an updated version of the rendered file: 
[04-differential_expression.nb.html.zip](https://github.com/AlexsLemonade/training-modules/files/10240246/04-differential_expression.nb.html.zip)

- I removed any filtering by genes and use all genes, not just HVG's now for differential expression. This fixed the plot of the dispersion estimates which looks better! There are just a lot of genes now, and only a few of the top genes are shown in the volcano plot because of overlaps but I think that's okay. 
- I also removed the plot showing the logFC shrinkage and removed it as a separate section, but did not remove the code since we still want to do that and mention it. 
- In the beginning when discussing the distribution of cell types across samples and subdiagnoses I replaced the UMAPs with a stacked bar chart. 
- I also removed some of the extra UMAPs that we probably don't need and included and added multiple features to the violin plot looking at gene expression across subdiagnoses and cell types. I tried adding 4 genes at first and it looked really crowded so I picked one upregulated gene and one downregulated gene. 

I did not suggest any genes to explore in the chunk that we let them explore in. Partially because I'm not quite sure what we should be defining as an interesting pattern... I think we should encourage them to look at some of the other top up and down regulated genes that are labeled in the volcano plot. Or alternatively we can remove the exploration because they will have the opportunity in the exercises to make plots? 

The other thing I did not do was make a ton of text changes around explanations. I plan on addressing those comments when cleaning up each section in subsequent PRs.

jashapiro commented on scRNA-seq-advanced/04-differential_expression.Rmd:
````diff
@@ -0,0 +1,508 @@
+)
+
+# subset sce to only contain samples with library IDs of interest 
+samples_to_keep <- integrated_sce$sample %in% library_ids
+rms_sce <- integrated_sce[,samples_to_keep]
+```
````
It looks like this notebook got killed in CI for memory usage. We probably want to get rid of the integrated object as soon as we subset it here. If that doesn't work, we may need to think about other solutions.

sjspielman commented on scRNA-seq-advanced/04-differential_expression.Rmd:
```diff
@@ -0,0 +1,511 @@
+                labSize = 3  # smaller labels
+                ) +
+  # change the overall theme
+  theme_classic() +
```
`theme_bw()` maybe? I think that will be more consistent with other notebooks.

sjspielman reviewed (approved):
This seems good to me for this stage of development and ready to move forward to next steps. I left some small comments, and assuming it still builds 👍 

One main organizational items I thought of (to bear in mind for followup PRs!) is that we do a lot of SCE wrangling here that goes back and forth between `DataFrame` and `data.frame`. I think an overall quick statement/explanation for why we often do this back and forth will be useful context for the notebook as a whole and as tips for working with SCEs more generally.
